homeassistant/components/romy/config_flow.py

```diff
@@ -0,0 +1,151 @@
+        self.romy_info: dict[str, Any] = {}
+
+    async def async_step_user(
+        self, user_input: dict[str, str] | None = None
+    ) -> FlowResult:
+        """Handle the user step."""
+        errors: dict[str, str] = {}
+        data = _schema_with_host()
+
+        if user_input:
+            self.romy_info[CONF_HOST] = user_input[CONF_HOST]
+
+            new_romy = await romy.create_romy(self.romy_info[CONF_HOST], "")
+
+            if not new_romy.is_initialized:
+                errors[CONF_HOST] = "cannot_connect"
+            else:
+                # check if already setuped
+                await self.async_set_unique_id(new_romy.unique_id)
+                self._abort_if_unique_id_configured()
+
+                self.romy_info[CONF_NAME] = new_romy.name
+
+                if not new_romy.is_unlocked:
+                    return await self.async_step_unlock_http_interface()
+
+            if not errors:
+                return await self._async_step_finish_config()
+
+        return self.async_show_form(step_id="user", data_schema=data, errors=errors)
+
+    async def async_step_unlock_http_interface(
+        self, user_input: dict[str, str] | None = None
+    ) -> FlowResult:
+        """Unlock the http interface with password if robot is locked."""
+        errors: dict[str, str] = {}
+        data = _schema_with_password()
+
+        if user_input:
+            self.romy_info[CONF_PASSWORD] = user_input[CONF_PASSWORD]
+            new_romy = await romy.create_romy(
+                self.romy_info[CONF_HOST], self.romy_info[CONF_PASSWORD]
+            )
+
+            if not new_romy.is_initialized or not new_romy.is_unlocked:
+                errors[CONF_PASSWORD] = "invalid_auth"
+
+            if not errors:
+                return await self._async_step_finish_config()
+
+        return self.async_show_form(
+            step_id="unlock_http_interface", data_schema=data, errors=errors
+        )
+
+    async def async_step_zeroconf(
+        self, discovery_info: zeroconf.ZeroconfServiceInfo
+    ) -> FlowResult:
+        """Handle zeroconf discovery."""
+
+        LOGGER.debug("Zeroconf discovery_info: %s", discovery_info)
+
+        # connect and gather information from your ROMY
+        host = discovery_info.host
+        LOGGER.debug("ZeroConf Host: %s", host)
+
+        new_discovered_romy = await romy.create_romy(host, "")
+
+        name = new_discovered_romy.name
+        LOGGER.debug("ZeroConf Name: %s", name)
+
+        # get unique id and stop discovery if robot is already added
+        unique_id = new_discovered_romy.unique_id
+        LOGGER.debug("ZeroConf Unique_id: %s", unique_id)
+        await self.async_set_unique_id(unique_id)
+        self._abort_if_unique_id_configured()
+
+        self.context.update(
+            {
+                "title_placeholders": {"name": f"{name} ({host} / {unique_id})"},
+                "configuration_url": f"http://{host}:{new_discovered_romy.port}",
```

Comment:
Where are these context variables used? (I have personally never seen this in a config flow before, so trying to understand)

Reply:
get used in the discovery

![romy_discovery2](https://github.com/home-assistant/core/assets/2132412/7b7b7d69-439e-4d5f-9ad6-0c5cc6a57be1)

![romy_discovery](https://github.com/home-assistant/core/assets/2132412/d45faef6-64bc-463a-82e1-0d54535ff230)


the configuration_url is used for "visit device" button

![romy_device](https://github.com/home-assistant/core/assets/2132412/02d4c410-8724-4f30-9b58-d8c7e2ed5344)


(afair i copied it from the octoprint addon)